### ISSUE TITLE
Inconsistent type enforcement for finalize argument in ground state solvers

### ISSUE THREAD

**Issue by 20akshay00** [QuantumKitHub/MPSKit.jl#207]
`DMRG` methods take in a keyword argument `finalize`, while `GradientGrassmann` take in `finalize!`, meant to act as a callback function at the end of each solver iteration. `DMRG` is [flexible](https://github.com/QuantumKitHub/MPSKit.jl/blob/283a62c3a4889e6e8ae3b384caac49418e8c6fad/src/algorithms/groundstate/dmrg.jl#L19) with the type of the argument, whereas `GradientGrassmann` [casts](https://github.com/QuantumKitHub/MPSKit.jl/blob/master/src/algorithms/groundstate/gradient_grassmann.jl#L28) it into a `Function`. 

This was encountered when trying to use a struct as a function so that it carries local state while performing the required finalizations.

**Comment by lkdvos** [comment]
Additionally, I'm not even sure the signatures of the finalizers are the same.
OptimKit uses `finalize!(x, f, g, numiter) -> x, f, g`, while DMRG requires `finalize(iter, state, H, envs) -> state, envs`

**Comment by 20akshay00** [comment]
Yes, I noticed that also, but thats just by virtue of DMRG working differently than Gradient methods whereas the casting as far as I see, isn't strictly necessary (?).

However, I did wonder if there was a scope to introduce a unified callback interface for these solvers, but wasn't sure if it made sense given how differently these methods work. At the very least, I think it makes sense for a user to have an easier way to quickly profile optimization schemes to choose the best one for the system at hand without having to dig through these differences. But I suppose a separate issue is in order for that.
